Fix IIIF V3 manifests so they validate

eliotjordan opened the pull request:
Closes [#6605](https://github.com/pulibrary/figgy/issues/6605)

- Audio, Video, Playlist, ScannedMap, and MapSet resources all generate 100% valid V3 manifests.
- ScannedMaps and MapSets now return V3 by default
- Added tests into V3 specs to test generated manifests against the V3 JSON schema.
- Disable downloads on a per-canvas level and remove outdated UV-specific disableUI service. Newer versions of UV and other viewers like clover don't use this property.

tpendragon commented on app/services/manifest_builder_v3/canvas_rendering_builder.rb:
```diff
@@ -45,7 +45,11 @@ def resource
       def downloadable?
-        @record.respond_to?(:resource) || @record.is_a?(FileSet)
+        if @record.respond_to?(:resource) || @record.is_a?(FileSet)
+          @record.resource.decorate.downloadable?
```
Can you explain the change in logic here? I think this would be quite expensive, re-fetching the parent for every FileSet.

eliotjordan replied:
We can't (once we upgrade UV or in the places where we use clover) use the custom UV disableUI service. Most of the download rendering are on the individual canvases so we need to check if we should build them or not. I'll push up a commit that's more efficient.

tpendragon commented on spec/services/manifest_builder/by_resource_type/manifest_builder_ephemera_spec.rb:
```diff
@@ -44,9 +44,9 @@
+    it "the root label displays the original title only because it is single valued" do
       output = manifest_builder.build
-      expect(output["label"]).to eq ["title", "transliterated"]
+      expect(output["label"]).to eq "title"
```
What's the impact on viewer display for this change? Anything?

eliotjordan replied:
No impact. The viewer only displays one title.

tpendragon commented on spec/services/manifest_builder/by_resource_type/manifest_builder_scanned_resource_spec.rb:
```diff
@@ -158,9 +158,9 @@ def logical_structure(file_set_id)
+    it "displays a root label with a single string as label is a single value in Prezi V2" do
       output = manifest_builder.build
-      expect(output["label"]).to eq ["title1", "title2"]
+      expect(output["label"]).to eq "title1"
```
Same question as above - what's the impact of this change? Did UV used to concat?

eliotjordan replied:
UV does not concat as far as I can tell.

tpendragon commented on spec/services/manifest_builder_v3/by_resource_type/manifest_builder_v3_audio_resource_spec.rb:
```diff
@@ -49,14 +51,36 @@
+        change_set_persister.save(change_set: change_set)
+        output = manifest_builder.build
+
+        expect(output["items"][0]["accompanyingCanvas"]["width"]).to eq 200
```
Does this accompanyingCanvas render in UV?

eliotjordan replied:
Hmm... It doesn't appear to in this version of UV. I guess we could add the posterCanvas back until we switch over to Clover to audio and video.

eliotjordan replied:
Added back posterCanvas.

tpendragon commented on spec/services/manifest_builder_v3/by_resource_type/manifest_builder_v3_audio_resource_spec.rb:
```diff
@@ -208,7 +209,7 @@
       stub_findingaid(pulfa_id: "C0652")
       stub_findingaid(pulfa_id: "C0652_c0377")
-      IngestArchivalMediaBagJob.perform_now(collection_component: "C0652", bag_path: bag_path, user: user, member_of_collection_ids: [collection.id.to_s])
+      IngestArchivalMediaBagJob.perform_now(collection_component: "C0652", bag_path: bag_path, user: user, member_of_collection_ids: [collection.id.to_s], visibility: "open")
```
Why was the visibility change necessary?

eliotjordan replied:
Not necessary. Fixed.

tpendragon commented on spec/services/manifest_builder_v3/by_resource_type/manifest_builder_v3_playlist_spec.rb:
```diff
@@ -79,9 +84,6 @@
         expect(output).not_to be_empty
 
-        expect(output).to include("rendering")
-        expect(output["rendering"]).to be_empty
```
How do we provide access to a canvas' full tiff download? Is the rendering somewhere else? Or do we still have canvas renderings and this just makes sure we don't have blank values?

tpendragon reviewed (changes requested):
A few questions. Have you tested the display in Figgy with these changes? If so, any notes to provide there, or should I kick this around on staging?

tpendragon commented on app/services/manifest_builder_v3/record_property_builder.rb:
```diff
@@ -7,17 +7,7 @@ class RecordPropertyBuilder < IIIFManifest::V3::ManifestBuilder::RecordPropertyB
-        manifest["service"] << {
-          "@context" => "http://universalviewer.io/context.json",
-          "profile" => "http://universalviewer.io/ui-extensions-profile",
-          "disableUI" => ["mediaDownload"]
```
So if we remove this, does the download pane still go away for things where download is disabled?

eliotjordan replied:
It does not go away, but says there are no downloads. If you want to make this happen going forward (once we upgrade UV), I think we'll have to somehow pass the [uv_config_downloads_disabled](https://github.com/pulibrary/figgy/blob/main/public/uv/uv_config_downloads_disabled.json) config that we have. Though I don't see anyplace where we are using it currently. I

tpendragon replied:
That's surprising, I wonder if it's the config disabling the downloads and we're fine here? I would have expected it to still show that like, "download a jpg" selection that uses the IIIF image API

tpendragon commented on app/services/manifest_builder_v3/canvas_rendering_builder.rb:
```diff
@@ -42,11 +42,11 @@ def resource
       def downloadable?
-        if @record.respond_to?(:resource) || @record.is_a?(FileSet)
-          @record.resource.decorate.downloadable?
+        if @record.respond_to?(:resource)
```
I'm not sure this logic is equivalent - before FileSets always returned true.

I honestly do not know what this downloadable check is.

eliotjordan replied:
I think the logic was flawed before. It checks if the parent resource has downloadable set to "public" and whether it is in a public state. Only then does it render the download links.

eliotjordan commented:
I refactored the download logic and added a [test that the parent is only retrieved once](https://github.com/pulibrary/figgy/pull/6625/files#diff-71a71f165382cc6651ee6287e55f5d67612a7c5dbb4317bc237b63dc6233c13aR94) to call the `downloadable?` method. I discovered some [sneaky recursion](https://github.com/pulibrary/figgy/blob/6210853e9b5d9391de1b26b72bfa7e326dac054c/app/services/manifest_builder_v3/root_node.rb#L113) that had to be worked around. The parent resource was being retrieved at least once for every FileSet. This should speed up V3 manifest generation all around.

tpendragon reviewed (dismissed):
Nice! I'm so happy that test uncovered some more optimizations. Good work!

tpendragon commented:
@eliotjordan Looks like we need some coverage.
